envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt

```diff
@@ -0,0 +1,147 @@
+    companion object {
+
+        @JvmField
+        @RegisterExtension
+        val consul = ConsulExtension()
+
+        @JvmField
+        @RegisterExtension
+        val envoyControl = EnvoyControlExtension(consul)
+
+        @JvmField
+        @RegisterExtension
+        val service = EchoServiceExtension()
+
+        @JvmField
+        @RegisterExtension
+        val envoy = EnvoyExtension(envoyControl, service, config = Xds)
+    }
+
+    override fun consul() = consul
+
+    override fun envoyControl() = envoyControl
+
+    override fun service() = service
+
+    override fun envoy() = envoy
+
+    override fun expectedGrpcConnectionsGaugeValues() = mapOf(
+        CDS to 1,
+        EDS to 2, // separate streams for consul and echo
+        LDS to 1,
+        RDS to 2, // default_routes
+        SDS to 0,
+        ADS to 0,
+        UNKNOWN to 0
+    )
+}
+
+class AdsMetricsDiscoveryServerCallbackTest : MetricsDiscoveryServerCallbacksTest {
+    companion object {
+
+        @JvmField
+        @RegisterExtension
+        val consul = ConsulExtension()
+
+        @JvmField
+        @RegisterExtension
+        val envoyControl = EnvoyControlExtension(consul)
+
+        @JvmField
+        @RegisterExtension
+        val service = EchoServiceExtension()
+
+        @JvmField
+        @RegisterExtension
+        val envoy = EnvoyExtension(envoyControl, service, config = Ads)
+    }
+
+    override fun consul() = consul
+
+    override fun envoyControl() = envoyControl
+
+    override fun service() = service
+
+    override fun envoy() = envoy
+
+    override fun expectedGrpcConnectionsGaugeValues() = mapOf(
+        CDS to 0,
+        EDS to 0,
+        LDS to 0,
+        RDS to 0,
+        SDS to 0,
+        ADS to 1, // all info is exchanged on one stream
+        UNKNOWN to 0
+    )
+}
+
+interface MetricsDiscoveryServerCallbacksTest {
+
+    fun consul(): ConsulExtension
+
+    fun envoyControl(): EnvoyControlExtension
+
+    fun service(): EchoServiceExtension
+
+    fun envoy(): EnvoyExtension
+
+    @Test
+    fun `should measure gRPC connections`() {
+        // given
+        val meterRegistry = envoyControl().app.bean(MeterRegistry::class.java)
+        consul().server.operations.registerService(service(), name = "echo")
+
+        // expect
+        untilAsserted {
+            expectedGrpcConnectionsGaugeValues().forEach { (type, value) ->
+                val metric = "grpc.connections.${type.name.toLowerCase()}"
+                assertThat(meterRegistry.find(metric).gauge()).isNotNull
+                assertThat(meterRegistry.get(metric).gauge().value().toInt()).isEqualTo(value)
+            }
+        }
+    }
+
+    fun expectedGrpcConnectionsGaugeValues(): Map<MetricsDiscoveryServerCallbacks.StreamType, Int>
+
+    @Test
+    fun `should measure gRPC requests`() {
+        // given
+        val meterRegistry = envoyControl().app.bean(MeterRegistry::class.java)
```

Comment:
That looks so hacky

Reply:
Also, this is more of a question regarding `EnvoyControlExtension` and `EnvoyControlTestApp`. Is there anything wrong with more inline reify magic from `EnvoyControlTestConfiguration` so `EnvoyControlTestApp` does not do that?

Reply:
What I can do is to hide meter registry extraction under `envoyControl().app.meterRegistry()` if you prefer :)

As for reify magic – that seemed more hacky and less explicit for me ;) maybe it's just my conservative Java background.